src/engine/client/cl_cgame.cpp

```diff
@@ -1474,74 +1447,65 @@ void CGameVM::QVMSyscall(int index, Util::Reader& reader, IPC::Channel& channel)
 				Key_SetCatcher(catcher);
 			});
 			break;
 
-		case CG_KEY_GETKEYNUMFORBINDS:
-			IPC::HandleMsg<Key::GetKeynumForBindsMsg>(channel, std::move(reader), [this] (int team, const std::vector<std::string>& binds, std::vector<std::vector<int>>& result) {
+		case CG_KEY_GETKEYSFORBINDS:
+			IPC::HandleMsg<Keyboard::GetKeysForBindsMsg>(channel, std::move(reader), [this] (int team, const std::vector<std::string>& binds, std::vector<std::vector<Keyboard::Key>>& result) {
 				for (const auto& bind : binds) {
-					result.push_back({});
-					for (int i = 1; i < Util::ordinal(keyNum_t::MAX_KEYS); i++) {
-						if (i >= 'A' && i <= 'Z') {
-							continue; // Skip since these are equivalent to lowercase letters.
-						}
-						char buffer[MAX_STRING_CHARS];
-
-						Key_GetBindingBuf(i, team, buffer, MAX_STRING_CHARS);
-						if (bind == buffer) {
-							result.back().push_back(i);
-							continue;
-						}
-					}
+					result.push_back(Keyboard::GetKeysBoundTo(team, bind));
 				}
 			});
 			break;
 
-		case CG_KEY_KEYNUMTOSTRINGBUF:
-			IPC::HandleMsg<Key::KeyNumToStringMsg>(channel, std::move(reader), [this] (int keynum, std::string& result) {
-				result = Keyboard::KeyToString(Keyboard::Key::FromLegacyInt(keynum));
+		case CG_KEY_GETCHARFORSCANCODE:
+			IPC::HandleMsg<Keyboard::GetCharForScancodeMsg>(channel, std::move(reader), [this] (int scancode, int& result) {
+				result = Keyboard::GetCharForScancode(scancode);
+				if (!result) {
+					// Not sure if this fallback is ever useful. Usually SDL falls back on QWERTY itself.
+					result = Keyboard::ScancodeToAscii(scancode);
+				}
 			});
 			break;
 
 		case CG_KEY_SETBINDING:
-			IPC::HandleMsg<Key::SetBindingMsg>(channel, std::move(reader), [this] (int keyNum, int team, std::string cmd) {
-				Keyboard::SetBinding(Keyboard::Key::FromLegacyInt(keyNum), team, std::move(cmd));
+			IPC::HandleMsg<Keyboard::SetBindingMsg>(channel, std::move(reader), [this] (Keyboard::Key key, int team, std::string cmd) {
+				Keyboard::SetBinding(key, team, std::move(cmd));
 			});
 			break;
 
 		case CG_KEY_CLEARCMDBUTTONS:
-			IPC::HandleMsg<Key::ClearCmdButtonsMsg>(channel, std::move(reader), [this] {
+			IPC::HandleMsg<Keyboard::ClearCmdButtonsMsg>(channel, std::move(reader), [this] {
 				CL_ClearCmdButtons();
 			});
 			break;
 
 		case CG_KEY_CLEARSTATES:
-			IPC::HandleMsg<Key::ClearStatesMsg>(channel, std::move(reader), [this] {
+			IPC::HandleMsg<Keyboard::ClearStatesMsg>(channel, std::move(reader), [this] {
 				Key_ClearStates();
 			});
 			break;
 
 		case CG_KEY_KEYSDOWN:
-			IPC::HandleMsg<Key::KeysDownMsg>(channel, std::move(reader), [this] (std::vector<int> keys, std::vector<int>& list) {
+			IPC::HandleMsg<Keyboard::KeysDownMsg>(channel, std::move(reader), [this] (std::vector<Keyboard::Key> keys, std::vector<bool>& list) {
 				list.reserve(keys.size());
-				for (unsigned i = 0; i < keys.size(); ++i)
+				for (Keyboard::Key key : keys)
```

Comment:
ditto

Reply:
`IsValid` is called in `IsDown` (though maybe `IsNonnull` or something like that would be a more accurate name).